sos/deviceauth.py

```diff
@@ -0,0 +1,291 @@
+# Copyright (C) 2023 Red Hat, Inc., Jose Castillo <jcastillo@redhat.com>
+
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+import logging
+import requests
+import time
+from datetime import datetime, timedelta
+import os
+from sos.keyring import Key, KeyNotFoundError
+import shutil
+
+DEVICE_AUTH_CLIENT_ID = "sos-tools"
+GRANT_TYPE_DEVICE_CODE = "urn:ietf:params:oauth:grant-type:device_code"
+DATETIME_FORMAT = "%Y-%m-%dT%H:%M:%SZ"
+
+logger = logging.getLogger("sos")
+
+
+def try_read_refresh_token():
+    """Try to read locally stored refresh token
+
+    Returns:
+        str: Returns ODIC refresh token if found otherwise None
+    """
+    try:
+        RHELKey = Key()
+        key_refresh_token = RHELKey.search('sos-tools_refresh_token')
+        key_username = RHELKey.search('sos-tools-user')
+    except KeyNotFoundError:
+        logger.info("Refresh token does not exist in keyring or is expired.")
+        return None
+    except Exception as e:
+        logger.error("Error encoutered while accessing keyring. {}".format(e))
+        return None
+    if key_username.data.decode() != os.getlogin():
+        return None
+
+    return key_refresh_token.data.decode()
+
+
+class AuthClass:
+    """
+    Device Authorization Class
+    """
+
+    def __init__(self):
+
+        self.client_identifier_url = "https://sso.redhat.com/auth/"\
+            "realms/redhat-external/protocol/openid-connect/auth/device"
+        self.token_endpoint = "https://sso.redhat.com/auth/realms/"\
+            "redhat-external/protocol/openid-connect/token"
+        self._access_token = None
+        self._access_expires_at = None
+        self._refresh_token = None
+        self._refresh_expires_at = None
+        self._refresh_expires_in = None
+        self._user_verification_url = None
+        self.__device_code = None
+        self.RHELKey = Key()
+
+        # Lets check first if we have keyctl installed so we can
+        # store the token in the keyring
+        if not shutil.which("keyctl"):
```

Comment:
We can use `is_executable` from `sos.utilities`? Well, practically there would not be a big difference, just if we have an internal routine for the same..?

Reply:
ack, fixed this and will force push soon.